File: backend/catalog-service/src/main/java/me/doruk/catalogservice/entity/Event.java
package me.doruk.catalogservice.entity;

import jakarta.persistence.*;
import lombok.*;
import jakarta.validation.constraints.*;

import java.math.BigDecimal;
import java.sql.Date;

@Entity
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Builder
@Table(name = "event")
public class Event {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  @Column(name = "id")
  private Long id;

  @Column(name = "name")
  @NotBlank(message = "Event name must not be blank")
  @Size(max = 255, message = "Event name must not exceed 255 characters")
  private String name;

  @Column(name = "total_capacity")
  @Positive(message = "Total capacity must be a positive number")
  private int totalCapacity;

  @Column(name = "remaining_capacity")
  @PositiveOrZero(message = "Remaining capacity must be zero or a positive number")
  private int remainingCapacity;

  @JoinColumn(name = "venue_id")
  @ManyToOne
  @NotNull(message = "Venue must not be null")
  private Venue venue;

  @Column(name = "ticket_price")
  @NotNull(message = "Ticket price must not be null")
  @DecimalMin(value = "0.0", inclusive = true, message = "Ticket price must be at least 0.0")
  private BigDecimal ticketPrice;

  @Column(name = "event_date")
  @NotNull(message = "Event date must not be null")
  private Date eventDate;

  @Column(name = "description")
  @Size(max = 1000, message = "Description must not exceed 1000 characters")
  private String description;

  @Column(name = "image_url")
  @Size(max = 512, message = "Image URL must not exceed 512 characters")
  private String imageUrl;

  @PrePersist
  public void prePersist() {
    if (remainingCapacity == 0) {
      remainingCapacity = totalCapacity;
    }
  }
}

File: backend/catalog-service/Entities/Event.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace CatalogService.Entities;

[Table("event")]
public class Event
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    [Column("id")]
    public long Id { get; set; }

    [Column("name")]
    [Required(ErrorMessage = "Event name must not be blank")]
    [StringLength(255, ErrorMessage = "Event name must not exceed 255 characters")]
    public string Name { get; set; } = null!;

    [Column("total_capacity")]
    [Range(1, int.MaxValue, ErrorMessage = "Total capacity must be a positive number")]
    public int TotalCapacity { get; set; }

    [Column("remaining_capacity")]
    [Range(0, int.MaxValue, ErrorMessage = "Remaining capacity must be zero or a positive number")]
    public int RemainingCapacity { get; set; }

    [Column("venue_id")]
    public long VenueId { get; set; }

    [ForeignKey(nameof(VenueId))]
    public Venue Venue { get; set; } = null!;
    
    [Column("ticket_price")]
    [Required(ErrorMessage = "Ticket price must not be null")]
    [Range(0.0, double.MaxValue, ErrorMessage = "Ticket price must be at least 0.0")]
    public decimal TicketPrice { get; set; }

    [Column("event_date")]
    [Required(ErrorMessage = "Event date must not be null")]
    public DateTime EventDate { get; set; }

    [Column("description")]
    [StringLength(1000, ErrorMessage = "Description must not exceed 1000 characters")]
    public string Description { get; set; } = null!;

    [Column("image_url")]
    [StringLength(512, ErrorMessage = "Image URL must not exceed 512 characters")]
    public string ImageUrl { get; set; } = null!;

    private Event() { }

    public Event(long id, string name, int totalCapacity, int remainingCapacity, decimal ticketPrice, DateTime eventDate, string description, string imageUrl)
    {
        Id = id;
        Name = name;
        TotalCapacity = totalCapacity;
        RemainingCapacity = remainingCapacity;
        TicketPrice = ticketPrice;
        EventDate = eventDate;
        Description = description;
        ImageUrl = imageUrl;

        PrePersist();
    }

    public void AssignVenue(Venue venue)
    {
        Venue = venue;
    }

    private void PrePersist()
    {
        if (RemainingCapacity == 0)
        {
            RemainingCapacity = TotalCapacity;
        }
    }
}
